### PR TITLE
[inductor][cpu] Fix max parallel depth to prevent OMP fork overhead when freezing

### LINKED ISSUE #190757: `TORCHINDUCTOR_FREEZING=1` makes CPU inductor generate a fused kernel ~15x slower than eager (shifted-window attention)
## Summary

On CPU with `torch.compile(backend="inductor")`, enabling `TORCHINDUCTOR_FREEZING=1`
produces a fused kernel that is dramatically **slower** than both eager and non-frozen
compile. Without freezing, compile is faster than eager as expected; turning freezing
on regresses the same graph by more than an order of magnitude.

All numbers below are **relative to eager** (median wall time of the forward pass),
measured on an **Intel Xeon 6** CPU. I'm not quoting absolute latencies, only ratios.

| mode | time vs eager |
|------|---------------|
| eager | 1.0x (baseline) |
| `compile` (no freezing) | ~0.5x (faster, expected) |
| `compile` + `TORCHINDUCTOR_FREEZING=1` | **~15x (much slower)** |

This is **not dtype-specific** — it reproduces with the same magnitude in fp16,
bf16 and fp32 (freezing lands at roughly the same absolute time in all three,
i.e. ~12-18x eager depending on dtype).

A profiler run shows ~97% of the time is spent inside a single generated fused
C++ kernel (`## Call CompiledFxGraph ##`); no individual `aten` op is slow. This
looks like a codegen/layout pathology introduced by the freezing pass, not a
slow operator. Outputs are numerically identical with and without freezing, so
it's purely a performance issue.

## Where it was found

Originally observed on the Hugging Face `transformers` model
[`laion/clap-htsat-fused`](https://huggingface.co/laion/clap-htsat-fused) — its
HTSAT audio tower (a Swin-style hierarchical transformer). On that real model the
audio tower under freezing is **~1.7x slower than eager**, while non-frozen
compile is ~0.6x. I then reduced it to a self-contained, dependency-free script
(below) where the effect is much larger (~15x), presumably because the reduced
graph is dominated by the offending pattern.

## Minimal reproduction (pure PyTorch, no dependencies)

By ablation, the regression needs the **interaction of three ingredients**;
removing any single one makes freezing as fast as (or faster than) plain compile:

1. a bicubic `F.interpolate` to a **tall / asymmetric** shape (`1024x64`) that is
   then reshaped to a square feature map — a square `256x256` interpolate does
   **not** trigger it;
2. a conv patch-embed followed by a **sigmoid residual gate** `g * sigmoid(conv(g))`;
3. **one shifted-window masked attention** layer — `torch.roll` plus a
   `masked_fill` attention bias added through a 5-D `view`.

Random weights are enough; this is not numerics-dependent.

```python
import math, os, sys, time
import torch
import torch.nn as nn
import torch.nn.functional as F

torch.manual_seed(0)
C, WS, SHIFT, NH = 96, 8, 4, 4


class Net(nn.Module):
    def __init__(self):
        super().__init__()
        self.proj = nn.Conv2d(1, C, 4, 4)
        self.gate = nn.Conv2d(C, C, 1)
        self.ln = nn.LayerNorm(C)
        self.qkv = nn.Linear(C, 3 * C)
        self.out = nn.Linear(C, C)

    def forward(self, x):
        x = F.interpolate(x, (1024, 64), mode="bicubic", align_corners=True).reshape(1, 1, 256, 256)
        g = self.proj(x)
        x = g * torch.sigmoid(self.gate(g))
        b, c, h, w = x.shape
        x = self.ln(x.flatten(2).transpose(1, 2)).view(b, h, w, c)
        x = torch.roll(x, (-SHIFT, -SHIFT), (1, 2))
        win = x.view(b, h // WS, WS, w // WS, WS, c).permute(0, 1, 3, 2, 4, 5).reshape(-1, WS * WS, c)
        n = win.shape[0]
        r = (torch.arange(h) >= h - WS).long() + (torch.arange(h) >= h - SHIFT).long()
        reg = (r[:, None] * 3 + r[None, :]).view(h // WS, WS, w // WS, WS).permute(0, 2, 1, 3).reshape(-1, WS * WS)
        mask = (reg.unsqueeze(1) - reg.unsqueeze(2)).half().masked_fill_(reg.unsqueeze(1) != reg.unsqueeze(2), -100.0)
        q, k, v = self.qkv(win).view(n, WS * WS, 3, NH, c // NH).permute(2, 0, 3, 1, 4)
        s = q @ k.transpose(-1, -2) / math.sqrt(c // NH)
        s = (s.view(1, n, NH, WS * WS, WS * WS) + mask.unsqueeze(1).unsqueeze(0)).view(-1, NH, WS * WS, WS * WS)
        o = (s.softmax(-1) @ v).transpose(1, 2).reshape(n, WS * WS, c)
        return self.out(o)


def bench(m, x, n=30):
    with torch.inference_mode():
        for _ in range(10):
            m(x)
        ts = []
        for _ in range(n):
            s = time.time()
            m(x)
            ts.append((time.time() - s) * 1000)
    return sorted(ts)[n // 2]


x = torch.randn(1, 1, 1001, 64, dtype=torch.float16)
m = Net().eval().to(torch.float16)
mode = sys.argv[1] if len(sys.argv) > 1 else "eager"
if mode == "compile":
    m.forward = torch.compile(m.forward, backend="inductor")
tag = "eager" if mode != "compile" else ("compile+freeze" if os.environ.get("TORCHINDUCTOR_FREEZING") == "1" else "compile")
print(f"{tag:16s} median={bench(m, x):.2f}ms")
```

Run (bind to one NUMA node for stable numbers):

```bash
numactl -N 0 -m 0 python repro.py                                # eager
numactl -N 0 -m 0 python repro.py compile                        # compile
TORCHINDUCTOR_FREEZING=1 numactl -N 0 -m 0 python repro.py compile   # compile + freezing
```

The third run is ~15x slower than the first; the second is faster than the first.

## Original model reproduction (via `transformers`)

For completeness, the same regression on the real model. This needs
`transformers`; the tower is loaded directly with random-free pretrained weights
(only the forward-pass shape/branch matters — `is_longer=True` selects the
4-channel fusion path that the feature extractor emits for this model):

```python
import os, sys, time, torch
from transformers import ClapModel

torch.manual_seed(0)
audio = ClapModel.from_pretrained("laion/clap-htsat-fused", dtype=torch.float16).eval().audio_model
inputs = dict(
    input_features=torch.randn(1, 4, 1001, 64, dtype=torch.float16),
    is_longer=torch.ones(1, 1, dtype=torch.bool),
)

if len(sys.argv) > 1 and sys.argv[1] == "compile":
    audio.forward = torch.compile(audio.forward, backend="inductor")

with torch.inference_mode():
    for _ in range(10):
        audio(**inputs)
    ts = []
    for _ in range(30):
        s = time.time()
        audio(**inputs)
        ts.append((time.time() - s) * 1000)

tag = "eager" if len(sys.argv) < 2 else ("compile+freeze" if os.environ.get("TORCHINDUCTOR_FREEZING") == "1" else "compile")
print(f"{tag:16s} median={sorted(ts)[15]:.2f}ms")
```

Here `compile` is ~0.6x eager, but `compile + TORCHINDUCTOR_FREEZING=1` is ~1.7x eager.

## Expected behavior

`TORCHINDUCTOR_FREEZING=1` should not make the compiled graph slower than eager
(and certainly not by an order of magnitude). At worst it should match non-frozen
compile.

## Environment

- CPU: Intel Xeon 6
- `torch.compile` backend: `inductor`, CPU
- dtype: reproduces in fp16, bf16 and fp32
- OS: Linux
- allocator: measured with `tcmalloc` preloaded; results are essentially the same without it

### Versions

torch 2.14.0.dev20260719+cpu

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by frost-intel**
Fixes #190757

LoopNest.max_parallel_depth decides whether to relocate parallelism from the outer loops onto an inner reduction loop by comparing the outer loops' work against the inner loop size. The outer-work estimate multiplies `FloorDiv(loop.size, loop.steps)` over the leading loops. When an outer loop is vectorized and its size is smaller than the vector step (e.g. size=4, steps=16 for a float SIMD width of 16), `FloorDiv(4, 16) == 0` zeroes the entire product. The heuristic then concludes the outer loops have ~0 work and moves parallelism onto the reduction, emitting a #pragma omp parallel region inside a serial outer loop — re-forking the thread pool on every iteration.

This changes `FloorDiv` to `CeilDiv` so a sub-vector-width loop contributes 1 iteration rather than underflowing to 0.

# Motivation
A model with `F.interpolate` to a tall/asymmetric shape + a conv patch-embed with a sigmoid gate + one shifted-window masked-attention layer regresses ~30-50x when compiled with `TORCHINDUCTOR_FREEZING=1` versus plain torch.compile, across different data types.

Root cause: freezing propagates a channels-last / head-innermost layout (from the mkldnn conv/linear rewrites) into the attention block, leaving the softmax reduction with a size-4 (num-heads) vectorized outer loop. The heuristic above then parallelized the 64-element reduction inside a serial 4096-iteration loop, so the kernel re-forked a 72-thread team ~4096× per call.

# Result
On the reporter's target model (freezing, 72 threads):
| *Time (ms)* | Eager | Compile | Compile+Freezing | Compile+Freezing+(This PR)
|--------|--------|--------|--------|---------|
| Small Repro | 2.07 | 1.37 | 82.87 | 0.95 |
| Full Model | 82.87 | 62.97 | 149.91 | 60.87 |

<details>
  <summary>Small Repro</summary>

```python
import math, os, sys, time
import torch
import torch.nn as nn
import torch.nn.functional as F

torch.manual_seed(0)
C, WS, SHIFT, NH = 96, 8, 4, 4


class Net(nn.Module):
    def __init__(self):
        super().__init__()
        self.proj = nn.Conv2d(1, C, 4, 4)
        self.gate = nn.Conv2d(C, C, 1)
        self.ln = nn.LayerNorm(C)
        self.qkv = nn.Linear(C, 3 * C)
        self.out = nn.Linear(C, C)

    def forward(self, x):
        x = F.interpolate(x, (1024, 64), mode="bicubic", align_corners=True).reshape(1, 1, 256, 256)
        g = self.proj(x)
        x = g * torch.sigmoid(self.gate(g))
        b, c, h, w = x.shape
        x = self.ln(x.flatten(2).transpose(1, 2)).view(b, h, w, c)
        x = torch.roll(x, (-SHIFT, -SHIFT), (1, 2))
        win = x.view(b, h // WS, WS, w // WS, WS, c).permute(0, 1, 3, 2, 4, 5).reshape(-1, WS * WS, c)
        n = win.shape[0]
        r = (torch.arange(h) >= h - WS).long() + (torch.arange(h) >= h - SHIFT).long()
        reg = (r[:, None] * 3 + r[None, :]).view(h // WS, WS, w // WS, WS).permute(0, 2, 1, 3).reshape(-1, WS * WS)
        mask = (reg.unsqueeze(1) - reg.unsqueeze(2)).half().masked_fill_(reg.unsqueeze(1) != reg.unsqueeze(2), -100.0)
        q, k, v = self.qkv(win).view(n, WS * WS, 3, NH, c // NH).permute(2, 0, 3, 1, 4)
        s = q @ k.transpose(-1, -2) / math.sqrt(c // NH)
        s = (s.view(1, n, NH, WS * WS, WS * WS) + mask.unsqueeze(1).unsqueeze(0)).view(-1, NH, WS * WS, WS * WS)
        o = (s.softmax(-1) @ v).transpose(1, 2).reshape(n, WS * WS, c)
        return self.out(o)


def bench(m, x, n=30):
    with torch.inference_mode():
        for _ in range(10):
            m(x)
        ts = []
        for _ in range(n):
            s = time.time()
            m(x)
            ts.append((time.time() - s) * 1000)
    return sorted(ts)[n // 2]


x = torch.randn(1, 1, 1001, 64, dtype=torch.float16)
m = Net().eval().to(torch.float16)
mode = sys.argv[1] if len(sys.argv) > 1 else "eager"
if mode == "compile":
    m.forward = torch.compile(m.forward, backend="inductor")
tag = "eager" if mode != "compile" else ("compile+freeze" if os.environ.get("TORCHINDUCTOR_FREEZING") == "1" else "compile")
print(f"{tag:16s} median={bench(m, x):.2f}ms")
```

</details>

<details>
  <summary>Full Model</summary>

```python
import os, sys, time, torch
from transformers import ClapModel

torch.manual_seed(0)
audio = ClapModel.from_pretrained("laion/clap-htsat-fused", dtype=torch.float16).eval().audio_model
inputs = dict(
    input_features=torch.randn(1, 4, 1001, 64, dtype=torch.float16),
    is_longer=torch.ones(1, 1, dtype=torch.bool),
)

if len(sys.argv) > 1 and sys.argv[1] == "compile":
    audio.forward = torch.compile(audio.forward, backend="inductor")

with torch.inference_mode():
    for _ in range(10):
        audio(**inputs)
    ts = []
    for _ in range(30):
        s = time.time()
        audio(**inputs)
        ts.append((time.time() - s) * 1000)

tag = "eager" if len(sys.argv) < 2 else ("compile+freeze" if os.environ.get("TORCHINDUCTOR_FREEZING") == "1" else "compile")
print(f"{tag:16s} median={sorted(ts)[15]:.2f}ms")
```

</details>

# Test
Adds `test_parallel_reduction_not_nested_in_loop` to `test/inductor/test_cpu_repro.py`, which compiles the pattern under `config.patch(freezing=True)` and asserts no #pragma omp parallel region is nested inside a loop body. Verified it fails without the fix and passes with it; the existing reduction/tiling/vec/parallel-reduction tests continue to pass (66 passed / 2 skipped).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by EikanWang** [comment]
@frost-intel , overall, it looks good to me. Could you please share the kernel snippets before and after the changes?

**Comment by frost-intel** [comment]
> @frost-intel , overall, it looks good to me. Could you please share the kernel snippets before and after the changes?

The before/after looks much simpler in that it doesn't use the partial accumulator array, etc, but that's not the primary issue here. The major problem is the inner `#pragam omp parallel num_threads(72)` which results in a new fork for each of the 4096 iterations.

### Before
```c
{
    #pragma GCC ivdep
    for(int64_t x0=0L; x0<4096L; x0+=1L)          // SERIAL outer loop
    {
        for(int64_t x1=0L; x1<4L; x1+=16L)
        {
            float tmp_acc0 = 0;
            at::vec::Vectorized<float> tmp_acc0_vec = at::vec::Vectorized<float>(0);
            at::vec::Vectorized<float> tmp_acc0_vec_arr[72];   // per-thread partials
            for (int i = 0; i < 72; i++) tmp_acc0_vec_arr[i] = at::vec::Vectorized<float>(0);
            float tmp_acc0_arr[72];
            for (int i = 0; i < 72; i++) tmp_acc0_arr[i] = 0;

            #pragma omp parallel num_threads(72)   // <-- team re-forked 4096x per call
            {
                int tid = omp_get_thread_num();
                at::vec::Vectorized<float> tmp_acc0_vec_local = at::vec::Vectorized<float>(0);
                float tmp_acc0_local = 0;
                #pragma omp for
                for(int64_t x2=0L; x2<64L; x2+=1L)          // reduce 64 across 72 threads
                {
                    if(C10_LIKELY(x1 >= 0L && x1 < 4L)) {
                        auto tmp0 = at::vec::Vectorized<float>::loadu(out_ptr1 + (x1 + 4L*x2 + 256L*x0), 4L);
                        tmp_acc0_vec_local = sum_masked_reduce(tmp_acc0_vec_local, tmp0, 4L);
                    }
                }
                tmp_acc0_vec_arr[tid] = tmp_acc0_vec_local;
                tmp_acc0_arr[tid] = tmp_acc0_local;
            }
            for (int tid = 0; tid < 72; tid++) tmp_acc0_vec = tmp_acc0_vec + tmp_acc0_vec_arr[tid];
            for (int tid = 0; tid < 72; tid++) tmp_acc0 = tmp_acc0 + tmp_acc0_arr[tid];
            if(C10_UNLIKELY(x1 >= 0L && x1 < 4L))
                tmp_acc0_vec.store(out_ptr2 + (x1 + 4L*x0), 4L);
        }
    }
}
```

### After
```c
{
    #pragma omp for
    for(int64_t x0=0L; x0<4096L; x0+=1L)          // outer loop parallelized once
    {
        for(int64_t x1=0L; x1<4L; x1+=16L)
        {
            float tmp_acc0 = 0;
            at::vec::Vectorized<float> tmp_acc0_vec = at::vec::Vectorized<float>(0);
            for(int64_t x2=0L; x2<64L; x2+=1L)         // serial vectorized inner reduction
            {
                if(C10_LIKELY(x1 >= 0L && x1 < 4L)) {
                    auto tmp0 = at::vec::Vectorized<float>::loadu(out_ptr1 + (x1 + 4L*x2 + 256L*x0), 4L);
                    tmp_acc0_vec = sum_masked_reduce(tmp_acc0_vec, tmp0, 4L);
                }
            }
            if(C10_UNLIKELY(x1 >= 0L && x1 < 4L))
                tmp_acc0_vec.store(out_ptr2 + (x1 + 4L*x0), 4L);
        }
    }
}
```

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
@claude review these changes again

**Review by jansel — CHANGES_REQUESTED** [review]
Fix lint errors

**Review by jansel — APPROVED** [review]
